developer/system/template-manager.md

```diff
@@ -0,0 +1,241 @@
+  - Note that newlines are converted to spaces, again for it to work better with existing tools.
+    This should not cause ambiguity as [the FreeDesktop specifications][freedesktop-spec] forbid spaces from the names of desktop files.
+
+[freedesktop-spec]: https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
+
+## Repository Management
+
+For UpdateVMs to access repository configuration, the package [qubes-repo-templates] is created with the following contents:
+
+- `/usr/share/qubes/repo-templates/*.repo`: repository configuration
+- `/usr/share/qubes/repo-templates/keys`: PGP keys
+
+As template keys may be less trusted, they are *not* added to the system RPM keychain but instead managed separately.
+
+[qubes-repo-templates]: https://github.com/WillyPillow/qubes-repo-templates
+
+## Qrexec Protocol
+
+Dom0 and management VMs without network access also need to interact with template repositories.
+The following qrexec calls that list and download templates are thus proposed.
+
+- `qubes.TemplateSearch`: wraps `dnf repoquery`
+- `qubes.TemplateDownload`: wraps `dnf download`
+
+### Input
+
+Both calls accept the following format from standard input:
+
+``` text
+arg1
+arg2
+...
+argN
+package-file-spec
+---
+repo config
+```
+
+In other words, the input consists of two parts separated by the line `---`.
+The first part contains some arguments and `package-file-spec` that indicates the pattern to be queried or downloaded.
+The following arguments are allowed:
+
+- `--enablerepo=<repoid>`
+- `--disablerepo=<repoid>`
+- `--repoid=<repoid>`
+- `--releasever=<release>`
+- `--refresh`
+
+where the usage is identical to that of DNF.
+
+For the exact definition of `package-file-spec`, refer to the DNF documentation.
+
+The second part contains the repository configurations in `yum.repos.d` format.
+
+### Output
+
+`qubes.TemplateSearch` prints each package in `%{name}|%{epoch}|%{version}|%{release}|%{reponame}|%{downloadsize}|%{buildtime}|%{license}|%{url}|%{summary}|%{description}|` format to standard output, separated by newlines.
+Note that there is a `|` at the end of the line.
+This is because `%{description}` may contain newlines, and doing so allows us to split the entries by `|\n`.
+(As we are using `dnf repoquery --qf`, we are unable to escape the newlines in advance.)
+
+`qubes.TemplateDownload`, on the other hand, directly outputs the downloaded content to standard output.
+
+## Machine-readable Output
+
+The commands `qvm-template list` and `qvm-template info` provide machine-readable output in both pipe(`|`)-separated and JSON format.
+See the `qvm-template` man page for details.
+
+## Interactions with Existing Tools
+
+### `qvm-remove`
+
+Should work straightforwardly.
```

Comment:
It's not clear to me whether this means "qvm-remove ignores all this" all "qvm-remove will remove templates properly, not leaving any gunk around" :) I think this needs a bit more explanation.